File: scripts/upload_words.py
import os
import requests
import cloudinary
from cloudinary import uploader

# רשימת המילים להעלאה
WORDS_TO_UPLOAD = [
    'Apple', 'Banana', 'Car', 'Dog', 'Cat', 'House', 'Tree', 'Sun', 'Moon', 'Star'
]
APP_TAG = 'english_kids_app'

def search_image_on_pixabay(query, api_key):
    url = f"https://pixabay.com/api/?key={api_key}&q={requests.utils.quote(query)}&image_type=photo&orientation=horizontal&per_page=3"
    print(f"  - Searching Pixabay for '{query}'...")
    response = requests.get(url)
    response.raise_for_status()
    data = response.json()
    if data.get('hits'):
        image_url = data['hits'][0]['webformatURL']
        print(f"  - Found image URL: {image_url}")
        return image_url
    print(f"  - No image found for '{query}'.")
    return None

def main():
    cloudinary_cloud_name = os.environ.get('CLOUDINARY_CLOUD_NAME')
    cloudinary_api_key = os.environ.get('CLOUDINARY_API_KEY')
    cloudinary_api_secret = os.environ.get('CLOUDINARY_API_SECRET')
    pixabay_api_key = os.environ.get('PIXABAY_API_KEY')

    if not all([pixabay_api_key, cloudinary_cloud_name, cloudinary_api_key, cloudinary_api_secret]):
        print("❌ Error: Missing required environment variables. Please set CLOUDINARY_CLOUD_NAME, CLOUDINARY_API_KEY, CLOUDINARY_API_SECRET, and PIXABAY_API_KEY.")
        return

    cloudinary.config(cloud_name=cloudinary_cloud_name, api_key=cloudinary_api_key, api_secret=cloudinary_api_secret)
    print("--- Python script started. Configured Cloudinary. ---")

    for word in WORDS_TO_UPLOAD:
        try:
            print(f"\nProcessing word: '{word}'")
            image_url = search_image_on_pixabay(word, pixabay_api_key)
            if not image_url: continue

            print(f"  - Uploading to Cloudinary with tags: ['{APP_TAG}', '{word.lower()}']")
            upload_result = uploader.upload(image_url, tags=[APP_TAG, word.lower()], public_id=word.lower(), folder=APP_TAG)

            print(f"  - ✅ Successfully uploaded '{word}'. Public ID: {upload_result.get('public_id')}")
        except Exception as e:
            print(f"  - ❌ An error occurred while processing '{word}': {e}")

    print("\n--- ✅ Python script finished. ---")

if __name__ == "__main__":
    main()

File: scripts/upload_words.dart
import 'dart:convert';
import 'dart:typed_data';
import 'package:http/http.dart' as http;
import 'package:firebase_core/firebase_core.dart';
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firebase_storage/firebase_storage.dart';
import 'package:english_learning_app/firebase_options.dart';
import 'package:english_learning_app/app_config.dart';

// --- רשימת המילים שברצונך להוסיף ---
const List<String> wordsToUpload = [
  'Apple',
  'Banana',
  'Car',
  'Dog',
  'Cat',
  'House',
  'Tree',
  'Sun',
  'Moon',
  'Star',
];

// ---- הפונקציה הראשית ----
void main() async {
  // אתחול שירותי Firebase
  await Firebase.initializeApp(options: DefaultFirebaseOptions.currentPlatform);

  final firestore = FirebaseFirestore.instance;
  final storage = FirebaseStorage.instance;
  final batch = firestore.batch();

  if (!AppConfig.hasFirebaseUserId) {
    print(
      '❌ Missing FIREBASE_USER_ID_FOR_UPLOAD. Provide it via --dart-define.',
    );
    return;
  }
  if (!AppConfig.hasPixabay) {
    print('❌ Missing PIXABAY_API_KEY. Provide it via --dart-define.');
    return;
  }

  final wordsCollection = firestore
      .collection('users')
      .doc(AppConfig.firebaseUserIdForUpload)
      .collection('words');

  print("Starting to process ${wordsToUpload.length} words...");

  for (String word in wordsToUpload) {
    try {
      print("\nProcessing word: '$word'");

      // 1. חפש תמונה ב-Pixabay
      final imageUrl = await searchImageOnPixabay(
        word,
        AppConfig.pixabayApiKey,
      );
      if (imageUrl == null) {
        print("  - Could not find image for '$word'. Skipping.");
        continue;
      }
      print("  - Found image URL: $imageUrl");

      // 2. הורד את התמונה מהאינטרנט
      final imageBytes = await downloadImage(imageUrl);
      print("  - Downloaded image successfully.");

      // 3. העלה את התמונה ל-Firebase Storage
      final storagePath = 'word_images/${word.toLowerCase()}.jpg';
      final finalImageUrl = await uploadImageToStorage(
        storage,
        imageBytes,
        storagePath,
      );
      print("  - Uploaded to Firebase Storage at: $finalImageUrl");

      // 4. הוסף את המידע ל-Firestore Batch
      final docRef = wordsCollection.doc();
      batch.set(docRef, {
        'word': word,
        'imageUrl': finalImageUrl,
        'isCompleted': false,
      });
      print("  - Added '$word' to the batch for Firestore.");
    } catch (e) {
      print("  - An error occurred while processing '$word': $e");
    }
  }

  // 5. בצע את כל הכתיבות ל-Firestore
  print("\nCommitting all words to Firestore...");
  await batch.commit();
  print("✅ Done! All words have been uploaded successfully.");
}

// ---- פונקציות עזר ----

// פונקציה לחיפוש תמונה ב-Pixabay
Future<String?> searchImageOnPixabay(String query, String apiKey) async {
  final url = Uri.parse(
    'https://pixabay.com/api/?key=$apiKey&q=${Uri.encodeComponent(query)}&image_type=illustration&orientation=horizontal&per_page=3',
  );
  final response = await http.get(url);

  if (response.statusCode == 200) {
    // --- נוסיף הדפסה של התשובה המלאה שקיבלנו ---
    print("  - Received a 200 OK response from Pixabay.");
    print("  - Response body: ${response.body}");

    final data = jsonDecode(response.body);
    if (data['hits'] != null && (data['hits'] as List).isNotEmpty) {
      // אם יש תוצאות, ניקח את כתובת התמונה הראשונה
      final imageUrl = data['hits'][0]['webformatURL'];
      return imageUrl;
    } else {
      // --- נוסיף הדפסה למקרה שאין תוצאות ---
      print("  - Pixabay found 0 images for '$query'.");
      return null;
    }
  } else {
    // --- נוסיף הדפסה למקרה של שגיאה ---
    print(
      "  - Pixabay request failed with status code: ${response.statusCode}",
    );
    print("  - Response body: ${response.body}");
    return null;
  }
}

// פונקציה להורדת התמונה מהכתובת שנמצאה
Future<List<int>> downloadImage(String url) async {
  final response = await http.get(Uri.parse(url));
  if (response.statusCode == 200) {
    return response.bodyBytes;
  }
  throw Exception('Failed to download image');
}

// פונקציה להעלאת התמונה ל-Firebase Storage וקבלת הקישור הסופי
Future<String> uploadImageToStorage(
  FirebaseStorage storage,
  List<int> imageBytes,
  String path,
) async {
  final ref = storage.ref().child(path);
  final uploadTask = ref.putData(Uint8List.fromList(imageBytes));
  final snapshot = await uploadTask.whenComplete(() => {});
  return await snapshot.ref.getDownloadURL();
}
